### PR TITLE
 List all functions that will be mocked with mock_call hint inside test files and test_<your_test_func> entrypoints #97 

### PR DESCRIPTION
**Pull request by jdembdr**
List all functions that will be mocked with mock_call hint inside test files and test_<your_test_func> entrypoints #97 
 

Time spent on this PR:1.5

# Pull Request type

Please check the type of change your PR introduces:

-   [ ] Bugfix
-   [x] Feature
-   [ ] Code style update (formatting, renaming)
-   [ ] Refactoring (no functional changes, no API changes)
-   [ ] Build-related changes
-   [ ] Documentation content changes
-   [ ] Testing
-   [ ] Other (please describe):

# What is the current behavior?

Issue Number: 97

# What is the new behavior?

///  let cairo_test = "func test_mock_call() {
///     let mock_ret_value = 42;
///     let func_to_mock = get_label_location(mocked_func);
///     %{ mock_call(func_to_mock, [mock_ret_value]) %}
///     %{ mock_call(name_of_my_func, [42, “shortstring”, ids.myVarName, Uint256 {high: 42, low: 0x0}]) %}
///     let x = mocked_func();
///     assert 42 = x;
///     return ();
/// }";
/// let res = extract_fname_mock_values(cairo_test).unwrap();
/// assert_eq!(res["test_mock_call"].len(), 2);
/// assert_eq!(res["test_mock_call"][0].fn_name, "func_to_mock");
/// assert_eq!(res["test_mock_call"][0].fn_args, "[mock_ret_value]");

# Does this introduce a breaking change?

-   [ ] Yes
-   [x] No

# Other information

### CODE REVIEW

File: src/cli/commands/test/mod.rs [hunk 1]
```diff
@@ -92,6 +94,111 @@ impl From<(String, TestStatus)> for TestResult {
 	}
 }
 
+/// structure used to store a Mocked function in a cairo test entrypoint
```
Comment (tdelabro) [inline src/cli/commands/test/mod.rs:97]: Put all of your new logic in a separated file, in the `io` module

File: src/cli/commands/test/mod.rs [hunk 2]
```diff
@@ -92,6 +94,111 @@ impl From<(String, TestStatus)> for TestResult {
...
+///     let x = mocked_func();
+///     assert 42 = x;
+///     return ();
+/// }";
+/// let res = extract_fname_mock_values(cairo_test).unwrap();
+/// assert_eq!(res["test_mock_call"].len(), 2);
+/// assert_eq!(res["test_mock_call"][0].fn_name, "func_to_mock");
+/// assert_eq!(res["test_mock_call"][0].fn_args, "[mock_ret_value]");
+/// ```
+pub fn extract_fname_mock_values(data: &str) -> Result<MockedFn, Error> {
```
Comment (tdelabro) [inline src/cli/commands/test/mod.rs:151]: rename data arg as something more expressive such as `function_code_block`

File: src/cli/commands/test/mod.rs [hunk 3]
```diff
@@ -92,6 +94,111 @@ impl From<(String, TestStatus)> for TestResult {
...
+/// ```ignore
+/// %{ mock_call(func_to_mock, [mock_ret_value]) %}
+/// ```
+const MOCK_RX: &str = r"%\{\s+mock_call\((?P<func_to_mock>.*),\s*(?P<mock_value>\[.*\])\)\s+%\}";
+
+/// compute the list of mocked functions in a text string
+/// Return a Result<MockedFn>
+///
+/// ```
+/// use cairo_foundry::cli::commands::test::extract_fname_mock_values;
```
Comment (tdelabro) [inline src/cli/commands/test/mod.rs:134]: https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#hiding-portions-of-the-example

You can hide this import

File: src/cli/commands/test/mod.rs [hunk 4]
```diff
@@ -92,6 +94,111 @@ impl From<(String, TestStatus)> for TestResult {
...
+
+	//extract fname and mock values
+	let mut i = 0;
+	while i != pos.len() - 1 {
+		match extract_fname_mock_values(&data[pos[i]..pos[i + 1]]) {
+			Ok(res) => {
+				println!("{res:?}");
+			},
+			_ => {
+				println!("FAILURE");
```
Comment (tdelabro) [inline src/cli/commands/test/mod.rs:194]: Return the error there rather than ignoring it

File: src/cli/commands/test/mod.rs [hunk 5]
```diff
@@ -92,6 +94,111 @@ impl From<(String, TestStatus)> for TestResult {
...
+	//prepare sections to parse, corresponding to functions body
+	let mut pos: Vec<usize> = fun_regex.find_iter(data.as_str()).map(|x| x.start()).collect();
+	// append the last line to parse entire file
+	if !pos.is_empty() {
+		pos.push(data.len());
+	}
+
+	//extract fname and mock values
+	let mut i = 0;
+	while i != pos.len() - 1 {
```
Comment (tdelabro) [inline src/cli/commands/test/mod.rs:188]: `for i in 0..pos.len()` would me more idiomatic

File: src/cli/commands/test/mod.rs [hunk 6]
```diff
@@ -92,6 +94,111 @@ impl From<(String, TestStatus)> for TestResult {
...
+/// fn_args: the list of parameters to replace the mock function call
+#[derive(Debug)]
+pub struct MockEntry<'a> {
+	pub fn_name: &'a str,
+	pub fn_args: &'a str,
+}
+
+/// subtype MockedFn store all the mocked functions in a single test entrypoint
+/// in a cairo file
+type MockedFn<'a> = HashMap<&'a str, Vec<MockEntry<'a>>>;
```
Comment (tdelabro) [inline src/cli/commands/test/mod.rs:108]: I'm not sure we need them organized by entrypoint.
The way we are going to use them is by reading the `compiled.json` and parse those functions data. We will do all of those at once.

File: src/cli/commands/test/mod.rs [hunk 7]
```diff
@@ -92,6 +94,111 @@ impl From<(String, TestStatus)> for TestResult {
 	}
 }
 
+/// structure used to store a Mocked function in a cairo test entrypoint
+/// fn_name: the name of the function to mock in the cairo file
+/// fn_args: the list of parameters to replace the mock function call
+#[derive(Debug)]
+pub struct MockEntry<'a> {
+	pub fn_name: &'a str,
+	pub fn_args: &'a str,
```
Comment (tdelabro) [inline src/cli/commands/test/mod.rs:103]: We don't need to know what the user intends to mock the call with at this point.
It's handle elsewhere in the execution

**Review by tdelabro — CHANGES_REQUESTED** [review]
Some lil changes to do. You also went a bit to far in term of what is needed.
+ we could use more tests. Maybe try the some edge case you can think of. Multiple tests in a file, multiple mock in a test.
How do you handle the same function being mocked multiple times ?
